### ISSUE TITLE
failed to do regsrv32 registration of MicrosoftInstrumentationEngine_x64.dll

### ISSUE THREAD

**Issue by sanikolov** [microsoft/CLRInstrumentationEngine#153]
Followed the build instructions and successfully generated all build artifacts.
Also ensured that all tests pass, namely 37 of 37, 42/42, 59/59 all pass.

In section [Environment Variables](https://github.com/microsoft/CLRInstrumentationEngine/blob/develop/docs/environment_variables.md), in the first occurrence of COR_PROFILER I see a bracketed UUID which implies that successful DLL registration is necessary, especially for framework versions < 4.

However, when I attempt DLL registration as shown below from an admin prompt, it fails with error 0x80070715, i.e.  "The specified resource type cannot be found in the image file."

```
C:\Windows\SysWOW64\regsvr32.exe /i C:\tmp\CLRInstrumentationEngine\bin\Debug\x64\InstrumentationEngine\x64\MicrosoftInstrumentationEngine_x64.dll
```

The above DLL is missing typical COM sections like REGISTRY, TYPELIB, String Table.
See the attached image.
![missing_com_sections](https://user-images.githubusercontent.com/19896827/67243649-0fa64400-f426-11e9-8855-2e2383035612.png)

Build tool version: vs2019 Professional 16.3.5
Host OS: Windows 10, 1903, build 18362.418

What am I missing here?
See also a [forward link](https://github.com/microsoft/CLRInstrumentationEngine/issues/57#issuecomment-541965597) to my previous message.

**Comment by delmyers** [comment]
@sanikolov It looks like you might be using the 32bit regsvr to try to register the 64bit dll. Have you tried running regsvr32 from C:\Windows\System32? The names of these folders are pretty cryptic on Windows. System32 is 64bit, and SysWOW64 is 32bit.

Note, if you are trying to use COM registration in order to get the CLR to load the ICorProfiler implementation, you will have to register both the 32 and 64bit dlls (each with the corresponding regsvr32.exe), otherwise you won't be able to profile both 32bit and 64bit applications.

Let us know how it goes for you. This is a little-tested scenario for us. All of our applications assume at least CLR 4, so there is no need for COM registration of the ICorProfiler. The COR_PROFILER_PATH* variables are typically preferred.

**Comment by sanikolov** [comment]
Thanks @delmyers for the quick reply. Tested registration from C:\Windows\System32\regsvr32 as well - made no difference. Still getting the same error code 0x80070715.

**Comment by delmyers** [comment]
@sanikolov, thanks for the follow-up

@WilliamXieMSFT, @wiktork, @jander-msft  it looks like we may have removed the implementation of DllInstall from the InstrumentationEngine/dllmain.cpp file. Does anyone remember why that might be the case?

**Comment by delmyers** [comment]
@sanikolov Can you try without "/i" I.e., just call:

``C:\Windows\System32\regsvr32.exe C:\tmp\CLRInstrumentationEngine\bin\Debug\x64\InstrumentationEngine\x64\MicrosoftInstrumentationEngine_x64.dll``

That will cause it to call DllRegisterServer without attempting to call DllInstall.

**Comment by sanikolov** [comment]
It is the call to DllRegisterServer that fails with code 0x80070715. DllInstall comes later, or in this case we never get around to it. I am nearly 100% sure that the cause of the error is missing COM sections from the DLL such as "TYPELIB", "REGISTRY", "String table". 2 or more of these sections should exist in every valid COM dll. See the image above where we see just 2 sections.

**Comment by delmyers** [comment]
@sanikolov You are probably correct. I just wanted to make sure that we have all of our bases covered.

My guess here is that we are depending on Atl's implementation of DllRegisterServer, but we aren't giving it the resources it needs to actually complete its operations.

Do you have a product that requires .NET < 4? Much of what happens inside ClrInstrumentationEngine requires .NET >= 4 (there may even be some places where calling into its functions will fail if there is a missing interface). If you do, that will help us to prioritize fixing this issue.

**Comment by sanikolov** [comment]
Yes, we have a shipping product that depends on framework versions 2 and 3, and would appreciate getting CLRInstrumentationEngine support for these versions.

**Comment by delmyers** [comment]
@sanikolov OK, I've marked this as a task/feature request. We have to do the work to support COM registration, and make sure that we can support .NET 3.5 at least. Like I said earlier, there may be some interfaces that we depend on that are not available in .NET 3.5.

If you would like to make a contribution to help achieve this, we would be open accepting a PR.
